vpr/src/place/initial_placement.cpp

```diff
@@ -236,101 +449,83 @@ static void initial_placement_pl_macros(int macros_max_num_tries, std::vector<st
+    }
 }
 
 /* Place blocks that are NOT a part of any macro.
  * We'll randomly place each block in the clustered netlist, one by one. */
-static void initial_placement_blocks(std::vector<std::vector<int>>& free_locations, enum e_pad_loc_type pad_loc_type, const std::vector<ClusterBlockId>& sorted_blocks) {
+static void place_block(int blocks_max_num_tries, ClusterBlockId blk_id, enum e_pad_loc_type pad_loc_type) {
     auto& cluster_ctx = g_vpr_ctx.clustering();
     auto& place_ctx = g_vpr_ctx.mutable_placement();
+    auto& floorplanning_ctx = g_vpr_ctx.floorplanning();
+    auto& device_ctx = g_vpr_ctx.device();
 
-    for (auto blk_id : sorted_blocks) {
-        bool placed = false;
-        while (!placed) {
-            /* -1 is a sentinel for a non-placed block, which the code in this routine will choose a location for.
-             * If the x value is not -1, we assume something else has already placed this block and we should leave it there.
-             * For example, if the user constrained it to a certain location, the block has already been placed.
-             */
-            if (place_ctx.block_locs[blk_id].loc.x != -1) {
-                //put VTR assert
-                placed = true;
-                continue;
-            }
-
-            auto logical_block = cluster_ctx.clb_nlist.block_type(blk_id);
-
-            /* Randomly select a free location of the appropriate type for blk_id.
-             * We have a linearized list of all the free locations that can
-             * accommodate a block of that type in free_locations[itype].
-             * Choose one randomly and put blk_id there. Then we don't want to pick
-             * that location again, so remove it from the free_locations array.
-             */
-
-            auto type = pick_placement_type(logical_block, 1, free_locations);
-
-            if (type == nullptr) {
-                VPR_FATAL_ERROR(VPR_ERROR_PLACE,
-                                "Initial placement failed.\n"
-                                "Could not place block %s (#%zu); no free locations of type %s (#%d).\n",
-                                cluster_ctx.clb_nlist.block_name(blk_id).c_str(), size_t(blk_id), logical_block->name, logical_block->index);
-            }
-
-            int itype = type->index;
-
-            auto possible_sub_tiles = get_possible_sub_tile_indices(type, logical_block);
-
-            int isub_tile = get_free_sub_tile(free_locations, itype, possible_sub_tiles);
+    bool placed = false;
+    while (!placed) {
+        /* -1 is a sentinel for a non-placed block, which the code in this routine will choose a location for.
+         * If the x value is not -1, we assume something else has already placed this block and we should leave it there.
+         * For example, if the user constrained it to a certain location, the block has already been placed.
+         */
+        if (place_ctx.block_locs[blk_id].loc.x != -1) {
```

Comment:
Wrap in a little function :  ! (block_placed(blk_id)).  block_placed return true unless .x is INVALID_X (which you can define to -1).